### PR TITLE
Worker Versioning

### PR DESCRIPTION
**Pull request by Sushisource**
Depends on https://github.com/temporalio/api/pull/293

## What was changed
Merge changes from the branch we've been developing worker versioning in. Most of these changes have already been reviewed independently.

## Why?
Implement the new feature!

## Checklist

1. Closes 

2. How was this tested:
Some added tests, more will likely be added in this repo after server changes have been merged - but there exist tests using this functionality in the server & features repos.

3. Any docs updates needed?
Yes, see https://github.com/temporalio/documentation/pull/1533

### CODE REVIEW

File: client/client.go
Comment (dnr) [inline client/client.go:193]: super-nitpicky but this might be easier to read?
```suggestion
	BuildIDOp_AddNewIDInNewDefaultSet = internal.UpdateBuildIDOp_NewSet
```

File: go.mod
Comment (dnr) [inline go.mod:15]: we need to update this after api lands?

File: internal/activity.go
Comment (dnr) [inline internal/activity.go:151]: use the same name for the value and type? otherwise it'll be hard to remember what it's called
```suggestion
		VersioningIntent VersioningIntent
```

File: internal/internal_task_handlers.go
Comment (dnr) [inline internal/internal_task_handlers.go:1593]: I see this logic repeated three times.. factor it out?

File: workflow/workflow_options.go
Comment (dnr) [inline workflow/workflow_options.go:81]: consistent naming
```suggestion
func WithVersioningIntent(ctx Context, intent temporal.VersioningIntent) Context {
```

File: go.mod
Comment (Sushisource) [reply go.mod:15]: Yes

File: internal/activity.go
Comment (Sushisource) [reply internal/activity.go:151]: I intentionally made them different so Go's idiotic docstring format would do the right thing. Otherwise you don't get a link to the type.

File: client/client.go
Comment (Sushisource) [reply client/client.go:193]: @cretz  preferred it this way. I don't care much either way

File: internal/error.go
Comment (bergundy) [inline internal/error.go:467]: Are we keeping this in the context or adding a new constructor for the CAN error?
I prefer the latter.

Comment (Sushisource) [reply internal/error.go:467]: I do too but that ship has sailed, we'll just keep it consistent here.

File: client/client.go
Comment (Sushisource) [reply client/client.go:193]: I think I just missed renaming the internal ones

Comment (cretz) [reply client/client.go]: Please don't use underscores. And ideally share a common prefix.

File: internal/internal_task_handlers.go
Comment (Sushisource) [reply internal/internal_task_handlers.go:1593]: To be clear the reason I didn't start it that way is that the defaults could, in principle, change on a per-command basis. But, if/when that ever happens then the shared thing can become un-shared so it's fine, changed this.

**Review by cretz — APPROVED** [review]
Minor stuff

File: internal/internal_task_handlers.go
Comment (bergundy) [inline internal/internal_task_handlers.go:1672]: We need to put the stamp on WFT failed too

File: internal/error.go
Comment (bergundy) [reply internal/error.go:467]: Manual construction works for me.

File: temporal/build_id_versioning.go
Comment (bergundy) [inline temporal/build_id_versioning.go:42]: Can we be consistent with the naming here?
Should be `UseCompatible` and `UseDefault` or drop the `Use` for both. I'd just drop `Use`.

File: internal/workflow_test.go
Comment (bergundy) [inline internal/workflow_test.go:59]: Can you add tests for the defaults for activity, child, and CAN please?
Can also be done in the features repo.

In any case let's get rid of this TODO before merging.

File: client/client.go
Comment (Sushisource) [reply client/client.go:193]: Yup, thanks

File: internal/internal_task_handlers.go
Comment (Sushisource) [reply internal/internal_task_handlers.go:1672]: It is already.

File: internal/workflow.go
Comment (Sushisource) [reply internal/workflow.go:1391]: Yes, it does, and no, because the others have a more sane pattern

File: temporal/build_id_versioning.go
Comment (Sushisource) [reply temporal/build_id_versioning.go:42]: Sure

File: internal/workflow_test.go
Comment (Sushisource) [reply internal/workflow_test.go:59]: This can't really be done in a super sane way in the unit tests that I can find easily, but, I've added a test for the helper function I made. Features repo can do the e2e validation

File: internal/workflow.go
Comment (bergundy) [reply internal/workflow.go:1391]: Please mark as experimental

File: internal/internal_task_handlers.go
Comment (bergundy) [reply internal/internal_task_handlers.go:1672]: Oh, I missed that

File: internal/workflow_test.go
Comment (bergundy) [reply internal/workflow_test.go]: SGTM

File: go.mod
Comment (bergundy) [inline go.mod:15]: Is this using api from master?
Seems like the version should have been bumped there.

Comment (Sushisource) [reply go.mod:15]: The version hasn't been bumped in quite some time. There's not really a meaningful reason to, our pattern for a bit has just been to let this sit this way.

**Review by bergundy — APPROVED** [review]
Already approving so this can be merged once CI passes and all comments are addressed.
@Sushisource will you also take care of sending only build ID if server has the required capability?

**Pull request merged**
